Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx

```diff
@@ -0,0 +1,361 @@
+  //_________________________________________________________________________________________________
+  void run(framework::ProcessingContext& pc)
+  {
+    if (mCCDBRequest) {
+      base::GRPGeomHelper::instance().checkUpdates(pc);
+    }
+
+    o2::globaltracking::RecoContainer recoData;
+    recoData.collectData(pc, *mDataRequest.get());
+
+    const auto& mchTracks = recoData.getMCHTracks();
+    const auto& mchmidMatches = recoData.getMCHMIDMatches();
+    const auto& mchClusters = recoData.getMCHTrackClusters();
+
+    for (auto const& mchmidMatch : mchmidMatches) {
+
+      int mchTrackID = mchmidMatch.getMCHRef().getIndex();
+      const auto& mchTrack = mchTracks[mchTrackID];
+      int first = mchTrack.getFirstClusterIdx();
+      int last = mchTrack.getLastClusterIdx();
+      int clIndex = -1;
+      mch::Track convertedTrack;
+
+      for (int i = first; i <= last; i++) {
+        clIndex += 1;
+        const auto& cluster = mchClusters[i];
+        mch::Cluster* mch_cluster = new mch::Cluster();
+        mch_cluster->x = cluster.x;
+        mch_cluster->y = cluster.y;
+        mch_cluster->z = cluster.z;
+
+        uint32_t ClUId = mch::Cluster::buildUniqueId(int(cluster.getDEId() / 100) - 1, cluster.getDEId(), clIndex);
+        mch_cluster->uid = ClUId;
+
+        mch_cluster->ex = cluster.ex;
+        mch_cluster->ey = cluster.ey;
+
+        convertedTrack.createParamAtCluster(*mch_cluster);
+      }
+
+      if (convertedTrack.getNClusters() > 9) {
+        // Erase removable track
+        if (!RemoveTrack(convertedTrack)) {
+          // mRecords.emplace_back(mAlign.ProcessTrack(convertedTrack, transformation, false, weightRecord));
+          pc.outputs().snapshot(Output{"MUON", "RECORD_MCHMID", 0}, mAlign.ProcessTrack(convertedTrack, transformation, false, weightRecord));
+        }
+      }
+    }
+    // pc.outputs().snapshot(Output{"MUON", "RECORD_MCHMID", 0}, mRecords);
+  }
+
+ private:
+  //_________________________________________________________________________________________________
+  bool RemoveTrack(mch::Track& track)
+  {
+    bool removeTrack = false;
+
+    try {
+      trackFitter.fit(track, false);
+    } catch (exception const& e) {
+      removeTrack = true;
+      return removeTrack;
+    }
+
+    auto itStartingParam = std::prev(track.rend());
+
+    while (true) {
+      try {
+        trackFitter.fit(track, true, false, (itStartingParam == track.rbegin()) ? nullptr : &itStartingParam);
+      } catch (exception const&) {
+        removeTrack = true;
+        break;
+      }
+
+      double worstLocalChi2 = -1.0;
+
+      track.tagRemovableClusters(0x1F, false);
+      auto itWorstParam = track.end();
+
+      for (auto itParam = track.begin(); itParam != track.end(); ++itParam) {
+        if (itParam->getLocalChi2() > worstLocalChi2) {
+          worstLocalChi2 = itParam->getLocalChi2();
+          itWorstParam = itParam;
+        }
+      }
+
+      if (worstLocalChi2 < trackFitter.getMaxChi2()) {
```

Comment:
I actually need this improvement for tracks, and what I made as change is to remove the manual setting for improve cut value so that it will be configured with `dpl-workflow.sh` with configurations key values like `MCHTracking.chamberResolutionX=0.4;MCHTracking.chamberResolutionY=0.4;MCHTracking.sigmaCutForTracking=7.;MCHDigitFilter.timeOffset=126;MCHTracking.sigmaCutForImprovement=6.;MCHTracking.maxCandidates=20000;`. With this, I think, the workflow will be able to get informed with those input parameters, am I right?

Reply:
Ok, I see, but `trackFitter.getMaxChi2()` is not what you want then. The parameter you need to set the chi2 cut is `sigmaCutForImprovement` and you can access it via `TrackerParam::Instance()`. You can have a look at the `TrackFinder` class, and in particular at the functions `init()`and `improveTracks()` for a concrete example.

Reply:
Ok, thanks a lot! I'm making modifications!

Reply:
By looking at your example of `TrackFinder`, I'm trying to do the same thing as you've done in `TrackFinder`; but there is one thing I'm not quite sure: if `trackFitter.setBendingVertexDispersion(trackerParam.bendingVertexDispersion);` is necessary for our case?

Reply:
This parameter is indeed used to initiate the refit. To be consistent it might be better to set it as well (although the default value in TrackFitter is equal to the default value in TrackerParam so it will not change the results if you keep the default).

Reply:
I see. Thanks a lot for explanation!